File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.py
class Solution:
    def minFallingPathSum(self, matrix: List[List[int]]) -> int:
        def fallingsum(i,j):
            if D.get((i,j))!=None:
                return D[(i,j)]+matrix[i][j]
            if i==len(matrix)-1:
                return matrix[i][j]
            else:
                a=b=math.inf
                if j>0:
                    a=fallingsum(i+1,j-1)
                if j<len(matrix[i])-1:
                    b=fallingsum(i+1,j+1)
                c=fallingsum(i+1,j)
                D[(i,j)]=min(a,b,c)
                return D[(i,j)]+matrix[i][j]  
        ans=math.inf
        D={}
        for i in range(len(matrix[0])):
            ans=min(ans,fallingsum(0,i))
        return ans

File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.cpp
class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& matrix) {
        int n=matrix.size();
        int m=matrix[0].size();
        for(int i=1;i<n;i++)
        {
            matrix[i][0]+=min(matrix[i-1][0],matrix[i-1][1]);
            matrix[i][m-1]+=min(matrix[i-1][m-1],matrix[i-1][m-2]);
            for(int j=1;j<m-1;j++)
            {
                matrix[i][j]+=min({matrix[i-1][j-1],matrix[i-1][j],matrix[i-1][j+1]});
            }
        }
        return *min_element(matrix[m-1].begin(),matrix[m-1].end());
    }
};